### ISSUE TITLE
thread 'tokio-runtime-worker' panicked at 'called `Result::unwrap()` on an `Err` value: could not fetch bootstrap

### ISSUE THREAD

**Issue by ltfschoen** [a16z/helios#254]
If this PR https://github.com/a16z/helios/pull/253/files, as at commit f4a7d4c66a770f437c79a0842d9b387544fce5de, if i run the following to build a Docker container using Ubuntu 23.04
```
./docker/docker.sh
```
and then after that script enters me into the container i run 
```
cargo run -p helios --example client
```
it outputs the following:
```
    Finished dev [unoptimized + debuginfo] target(s) in 0.65s
     Running `target/debug/examples/client`
Constructed client!
thread 'tokio-runtime-worker' panicked at 'called `Result::unwrap()` on an `Err` value: could not fetch bootstrap

Location:
    /opt/consensus/src/consensus.rs:304:26', /opt/consensus/src/consensus.rs:80:32
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
```

i'm using version `nightly-2023-07-31` in the Docker container, because the latest master branch uses that version in the rust-toolchain file

**Comment by ncitron** [comment]
In your example, you are using lightclientdata.org, which serves ethereum mainnet data, but you are using the goerli configs and a goerli execution rpc.

Also note that we are working on a couple of PRs to clean a few things up, so the current state of master will be in a bit of a weird spot for the next few days.

**Comment by ltfschoen** [comment]
actually, now that i'm connecting to Goerli with the right endpoint, that error `could not fetch bootstrap` is because it can't get the checkpoint 

i also noticed the instructions in config.md, which say to paste the following in ~/.helios/helios.toml.
so i ran the following
```
mkdir -p ~/.helios
mkdir -p /root/.helios/data/goerli
vim ~/.helios/helios.toml
```

Then pasted the following in ~/.helios/helios.toml
```
[goerli]
# The consensus rpc to use. This should be a trusted rpc endpoint. Defaults to Nimbus testnet.
consensus_rpc = "http://testing.prater.beacon-api.nimbus.team"
# [REQUIRED] The execution rpc to use. This should be a trusted rpc endpoint.
execution_rpc = "https://ethereum-goerli-rpc.allthatnode.com"
# The port to run the JSON-RPC server on. By default, Helios will use port 8545.
rpc_port = 8545
# The latest checkpoint. This should be a trusted checkpoint that is no greater than ~2 weeks old.
# If you are unsure what checkpoint to use, you can skip this option and set either `load_external_fallback` or `fallback` values (described below) to fetch a checkpoint. Though this is not recommended and less secure.
checkpoint = "0xb5c375696913865d7c0e166d87bc7c772b6210dc9edf149f4c7ddc6da0dd4495"
# The directory to store the checkpoint database in. If not provided, Helios will use "~/.helios/data/goerli", where `goerli` is the network.
# It is recommended to set this directory to a persistent location mapped to a fast storage device.
data_dir = "/root/.helios/data/goerli"
# The maximum age of a checkpoint in seconds. If the checkpoint is older than this, Helios will attempt to fetch a new checkpoint.
max_checkpoint_age = 86400
# A checkpoint fallback is used if no checkpoint is provided or the given checkpoint is too old.
# This is expected to be a trusted checkpoint sync api (like provided in https://github.com/ethpandaops/checkpoint-sync-health-checks/blob/master/_data/endpoints.yaml).
fallback = "https://sync-goerli.beaconcha.in"
# If no checkpoint is provided, or the checkpoint is too old, Helios will attempt to dynamically fetch a checkpoint from a maintained list of checkpoint sync apis.
# NOTE: This is an insecure feature and not recommended for production use. Checkpoint manipulation is possible.
load_external_fallback = true
```

But when i ran `cargo run -p helios --example client` i still got the same error `could not fetch bootstrap`

i'm also confused about how to check if the checkpoint hash is valid or not. my understanding is that a checkpoint hash should be a block hash for that network that is no greater than ~2 weeks old
if i go to https://goerli.etherscan.io/block/0xb5c375696913865d7c0e166d87bc7c772b6210dc9edf149f4c7ddc6da0dd4495 it says it can't locate that block hash. but that's a value that is provided in the default config.md.

and if i follow the instructions in config.md, by using a trusted checkpoint that is no greater than ~2 weeks old. such as block 9528294 that i found has block hash 0xfbec233115dc7497df5f6d06ca819a0de14c1d47a2b8903ab85607c0af4b5ff2, and that that block hash exists and was finalised https://goerli.etherscan.io/block/0xfbec233115dc7497df5f6d06ca819a0de14c1d47a2b8903ab85607c0af4b5ff2, so i add that in ~/.helios/helios.toml and also in client.rs, but when i run `cargo run -p helios --example client` i still get the same error

**Comment by ncitron** [comment]
Yes so the checkpoint is a beaconchain blockhash, rather than a execution block hash. This checkpoint is way to old (nearly a year). Normally, when it fails the fallback would kick in and automatically fetch a better one, but due to the refactors I mentioned that feature is currently not working.

To find a better checkpoint, you can go to https://prater.beaconcha.in/ and get the blockhash for the first block in any epoch. For example, the first blockhash in epoch 197110 is `0x7beab8f82587b1e9f2079beddebde49c2ed5c0da4ce86ea22de6a6b2dc7aa86b` which should work.

**Comment by ltfschoen** [comment]
> Yes so the checkpoint is a beaconchain blockhash, rather than a execution block hash. This checkpoint is way to old (nearly a year). Normally, when it fails the fallback would kick in and automatically fetch a better one, but due to the refactors I mentioned that feature is currently not working.
> 
> To find a better checkpoint, you can go to https://prater.beaconcha.in/ and get the blockhash for the first block in any epoch. For example, the first blockhash in epoch 197110 is `0x7beab8f82587b1e9f2079beddebde49c2ed5c0da4ce86ea22de6a6b2dc7aa86b` which should work.

thanks i tried using that blockhash from the beaconchain in this commit https://github.com/a16z/helios/pull/253/commits/b638ea602000adf62eac3b01f6ad8316945593d0 where i also added that blockhash to in the file ~/.helios/helios.toml as the `checkpoint` value for `[goerli]`, but for some reason i still get the same error still

**Comment by ncitron** [comment]
Hmm when I try to run it with these settings outside of docker by doing `cargo run -- -n goerli -c http://testing.prater.beacon-api.nimbus.team -e https://ethereum-goerli-rpc.allthatnode.com --checkpoint 0x7beab8f82587b1e9f2079beddebde49c2ed5c0da4ce86ea22de6a6b2dc7aa86b` it seems to sync correctly. Is it possible your docker environment is causing issues?

**Comment by ltfschoen** [comment]
> cargo run -- -n goerli -c http://testing.prater.beacon-api.nimbus.team -e https://ethereum-goerli-rpc.allthatnode.com --checkpoint 0x7beab8f82587b1e9f2079beddebde49c2ed5c0da4ce86ea22de6a6b2dc7aa86b

i just tried running the same command inside the Docker container:
```
cargo run -- -n goerli -c http://testing.prater.beacon-api.nimbus.team -e https://ethereum-goerli-rpc.allthatnode.com --checkpoint 0x7beab8f82587b1e9f2079beddebde49c2ed5c0da4ce86ea22de6a6b2dc7aa86b
```

and it output the 

```bash
[2023-08-16T21:43:20Z INFO  client::rpc] rpc server started at 127.0.0.1:8545
[2023-08-16T21:43:31Z INFO  consensus::consensus] updated head               slot=6307990  confidence=82.81%  age=00:05:05:31
[2023-08-16T21:43:31Z INFO  consensus::consensus] sync committee updated
[2023-08-16T21:43:31Z INFO  consensus::consensus] finalized slot             slot=6307904  confidence=82.81%  age=00:05:22:43
[2023-08-16T21:43:33Z INFO  consensus::consensus] updated head               slot=6309516  confidence=82.62%  age=00:00:00:21
[2023-08-16T21:43:33Z INFO  consensus::consensus] finalized slot             slot=6309440  confidence=82.62%  age=00:00:15:33
[2023-08-16T21:43:34Z INFO  consensus::consensus] consensus client in sync with checkpoint: 0x7beab8f82587b1e9f2079beddebde49c2ed5c0da4ce86ea22de6a6b2dc7aa86b
[2023-08-16T21:43:43Z INFO  consensus::consensus] updated head               slot=6309517  confidence=82.81%  age=00:00:00:19
```

**Comment by ltfschoen** [comment]
so things to try or consider now are
1) try change `load_external_fallback` to `false` in ~/.helios/helios.toml  --> didn't work
2) whether i need to use a proxy like in helios-ts where they use `executionRpc:"http://localhost:9001/proxy",`, and `consensusRpc: "http://localhost:9002/proxy",` but i'm only exposing port 8545
3) what other Helios CLI options to use from the help menu `cargo run -- --help` 
```
    -b, --rpc-bind-ip <RPC_BIND_IP>        [env: RPC_BIND_IP=]
    -c, --consensus-rpc <CONSENSUS_RPC>    [env: CONSENSUS_RPC=]
    -d, --data-dir <DATA_DIR>              [env: DATA_DIR=]
    -e, --execution-rpc <EXECUTION_RPC>    [env: EXECUTION_RPC=]
    -f, --fallback <FALLBACK>              [env: FALLBACK=]
    -h, --help                             Print help information
    -l, --load-external-fallback           [env: LOAD_EXTERNAL_FALLBACK=]
    -n, --network <NETWORK>                [default: mainnet]
    -p, --rpc-port <RPC_PORT>              [env: RPC_PORT=]
    -s, --strict-checkpoint-age            [env: STRICT_CHECKPOINT_AGE=]
    -V, --version                          Print version information
    -w, --checkpoint <CHECKPOINT>          [env: CHECKPOINT=]
```
so i tried
```
cargo run -p helios --example client -- -n goerli -c http://testing.prater.beacon-api.nimbus.team -e https://ethereum-goerli-rpc.allthatnode.com --checkpoint 0x7beab8f82587b1e9f2079beddebde49c2ed5c0da4ce86ea22de6a6b2dc7aa86b -p 8545 -l false -f https://sync-goerli.beaconcha.in -d /root/.helios/data/goerli -b 127.0.0.1
```
but that gave the same error too

**Comment by ncitron** [comment]
When running Helios as a library config files and cli arguments are not used. All configuration is done using the `ClientBuilder`

**Comment by ltfschoen** [comment]
so in consensus.rs, in the `Inner` function `bootstrap`, when it calls `get_bootstrap(&self.initial_checkpoint)`, that calls `get_bootstrap` in file nimbus_rpc.rs, snippet below
```
        let mut bootstrap = self
            .rpc
            .get_bootstrap(&self.initial_checkpoint)
            .await
            .map_err(|_| eyre!("could not fetch bootstrap"))?;
```
which constructs a `req` variable `"http://testing.prater.beacon-api.nimbus.team/eth/v1/beacon/light_client/bootstrap/0x7beab8f82587b1e9f2079beddebde49c2ed5c0da4ce86ea22de6a6b2dc7aa86b"`, which i can successfully load here http://testing.prater.beacon-api.nimbus.team/eth/v1/beacon/light_client/bootstrap/0x7beab8f82587b1e9f2079beddebde49c2ed5c0da4ce86ea22de6a6b2dc7aa86b
which shows it uses "capella" fork of Goerli, which starts from epoch 162304 according to helios/config/src/networks.rs

but it doesn't get past this line even though its using that `req` value, so it propagates an error from that function instead of returning `Ok(res.data)`
```
        let res = client
            .get(req)
            .send()
            .await
            .map_err(|e| RpcError::new("bootstrap", e))?
            .json::<BootstrapResponse>()
            .await
            .map_err(|e| RpcError::new("bootstrap", e))?;
```

**Comment by ltfschoen** [comment]
i changed the line to `.map_err(|e| eyre!("could not fetch bootstrap {}", e))?;` to output the error, and the error was:
```
rpc error on method: bootstrap, message: error sending request for url (http://testing.prater.beacon-api.nimbus.team/eth/v1/beacon/light_client/bootstrap/0x7beab8f82587b1e9f2079beddebde49c2ed5c0da4ce86ea22de6a6b2dc7aa86b): error trying to connect: dns error: task 8 was cancelled
```

i'm not sure yet how to debug further to find out what `task 8` refers to

**Comment by ltfschoen** [comment]
> task 8 was cancelled

so that error implementation is here https://github.com/tokio-rs/tokio/blob/master/tokio/src/runtime/task/error.rs#L138

**Comment by ltfschoen** [comment]
i get the similar errors running the **basic** example, as shown below

run basic example
```
cargo run -p helios --example basic
```

outputs error
```
[2023-08-17T05:50:56Z INFO  basic] Built client on network "GOERLI" with external checkpoint fallbacks
thread 'tokio-runtime-worker' panicked at 'called `Result::unwrap()` on an `Err` value: could not fetch bootstrap rpc error on method: bootstrap, message: error sending request for url (http://testing.prater.beacon-api.nimbus.team/eth/v1/beacon/light_client/bootstrap/0x428474ba0c1ce5049efe4324d503f69f0370fae2095954db4864c57097c90574): error trying to connect: dns error: task 9 was cancelled

Location:
    /opt/consensus/src/consensus.rs:308:26', /opt/consensus/src/consensus.rs:83:32
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
Error: block not available: latest
```

**Comment by ltfschoen** [comment]
i also get similar errors running the **call** example, as shown below

run **call** example

```
cargo run -p helios --example call
```

outputs error
```
[2023-08-17T05:46:31Z INFO  call] [ABIGEN] 0x8bb9a8baeec177ae55ac410c429cbbbbb9198cac::renderBroker(uint256) -> Response Length: 106476
thread 'tokio-runtime-worker' panicked at 'called `Result::unwrap()` on an `Err` value: could not fetch bootstrap rpc error on method: bootstrap, message: error decoding response body: missing field `data` at line 1 column 49

Location:
    /opt/consensus/src/consensus.rs:308:26', /opt/consensus/src/consensus.rs:83:32
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
[2023-08-17T05:46:32Z INFO  call] [ABIGEN] 0x8bb9a8baeec177ae55ac410c429cbbbbb9198cac::renderBroker(uint256, uint256) -> Response Length: 64549
[2023-08-17T05:46:32Z DEBUG call] Calling helios client on block: Latest
Error: block not available: latest

Location:
    /rustc/a17c7968b727d8413801961fc4e89869b6ab00d3/library/core/src/convert/mod.rs:716:9
```

**Comment by ltfschoen** [comment]
I think errors like `thread 'tokio-runtime-worker' panicked at 'called `Result::unwrap()` on an `Err` value: could not fetch bootstrap rpc error on method: bootstrap, message: error decoding response body: missing field `data` at line 1 column 49`  are caused because I switched from using Alchemy to a different provider, not recognising the warning that is already in the README.md file that says that the Execution RPC must be an Ethereum provider that supports the `eth_getProof` endpoint, and that providers like Infura does not currently support this, and that Alchemy is recommended

**Comment by ltfschoen** [comment]
> I think errors like `thread 'tokio-runtime-worker' panicked at 'called Result::unwrap() on an Err value: could not fetch bootstrap rpc error on method: bootstrap, message: error decoding response body: missing field data at line 1 column 49` are caused because I switched from using Alchemy to a different provider, not recognising the warning that is already in the README.md file that says that the Execution RPC must be an Ethereum provider that supports the `eth_getProof` endpoint, and that providers like Infura does not currently support this, and that Alchemy is recommended.
> 
> For example, I changed `GOERLI_EXECUTION_RPC=https://eth-goerli.g.alchemy.com/v2/XXXXX` in the .env.example file to `GOERLI_EXECUTION_RPC=https://ethereum-goerli-rpc.allthatnode.com`, but i think that causes the issue

actually i still seem to get the same error using Alchemy

**Comment by ncitron** [comment]
The most recent errors have been resolved as per #258
